### ISSUE TITLE
Demand saving controller

### ISSUE THREAD

**Issue by snorfalorpagus** [pywr/pywr#143]
We need a generic approach to demand saving. Previously this has been done in a "controller" object.

- The demand level should be calculated only once
- This is usually based on reservoir volume(s) in relation to a control curve(s)
- It would be useful to be able to have an aggregate demand saving level, based on the worst-case demand saving level of two or more reservoirs.
- We can then wrap the demand max_flow parameter in a demand saving adjustment parameter.

Need to be able to define this using JSON.

**Comment by snorfalorpagus** [comment]
The same controller could also be used to trigger changes to an MRF value in drought conditions.

I think we probably need to have a new `"controllers"` section in the JSON.

**Comment by jetuk** [comment]
We could add a new property to `Reservoir`, similar to `level` that can take a parameter but is intended to return the demand saving level? Or is this specialising the class too much?

Why can't we have a single controller `Parameter` in the `"parameters"` section that is reused by many nodes? It would have to do its calculations in `before` rather than `value` to be most efficient.

**Comment by snorfalorpagus** [comment]
I don't think calculating the demand saving level directly on a reservoir is the right way to go. Often the demand saving will be based on multiple reservoirs. Sometimes it's not based on a reservoir at all (e.g. for WRZs that have no storage, demand saving might be based on river flows).

Putting it in the `"parameters"` section is a good idea. Same for `before`.

We also need to be able to record the demand saving level. Do we need a recorder that takes a parameter directly, rather than a node?

**Comment by jetuk** [comment]
Here's how I would do this

A new subclass of `BaseControlCurveParameter` something like this. It assumes that self.control_curve is a `Parameter` that returns an iterable of multiple curves and then returns the first failure.
```
class MultiControlCurvePosition(BaseControlCurveParameter):

    def value(self, ts, scenario_index):
        i = scenario_index.global_id
        control_curves = self.control_curve.value(ts, scenario_index)
        node = self.node if self.storage_node is None else self.storage_node
        # Assumes control_curves is sorted highest to lowest

        for j, cc in enumerate(control_curves):
            if node.current_pc[i] >= cc:
                return j
        return len(control_curves)
```
You would then use this with `MaximumParameterCollection` to pass a number of them for different reservoirs and return the worst. 

Finally you have a `MultiLevelMonthlyScaledParameter` (or some better name) that is similar to `MonthlyProfileParameter` but has different factors for n levels. `MonthlyProfileParameter` should have have a scaling factor (i.e. the base demand) built in. All this could be done in `MonthlyProfileParameter` if the `level_param` was a keyword with `None` and ignored it.
```
class MultiLevelMonthlyScaledParameter(ConstantParameter):
    def __init__(self, factors, , level_param, scale=1.0, **kwargs):
        super(MultiLevelScaledParameter, self).__init__(scale=scale, **kwargs)
        self.factors = np.array(factors)  # this should be shape = (nlevels, 12)
        self.level_param = level_param  # this is returns the current level
    def value(self, ts, scenario_index):
        ilvl = self.level_param.value(ts, scenario_index)
        return self.factors[i, ts.datetime.month-1]*self.scale
```

**Comment by snorfalorpagus** [comment]
My first attempt looks like this:

```
class DemandSaving(Parameter):
    def __init__(self, model, factors, *args, **kwargs):
        super(DemandSaving, self).__init__(*args, **kwargs)
        self.factors = factors
        self.model = model
    
    def level(self, ts, si):
        raise NotImplementedError("DemandSaving must be subclassed!")
    
    def factor(self, ts, si):
        return self.factors[self.level(ts, si)]
parameter_registry.add(DemandSaving)
    
class DemandSavingReservoir(DemandSaving):
    def __init__(self, model, reservoirs, curves, *args, **kwargs):
        super(DemandSavingReservoir, self).__init__(model, *args, **kwargs)
        self.reservoirs = reservoirs
        self.curves = curves

    def level(self, ts, si):
        # calculate the aggregate reservoir volume as a factor
        max_volume = np.sum([storage.max_volume for storage in self.reservoirs])
        current_volume = np.sum([storage.volume[si.global_id] for storage in self.reservoirs])
        current_percent = current_volume / max_volume
        # calculate the demand saving level
        level = None
        for n, curve in enumerate(self.curves):
            if current_percent > curve.value(ts, si):
                level = n
                break
        if level is None:
            level = len(self.curves)
        return level
parameter_registry.add(DemandSavingReservoir)

class DemandParameter(Parameter):
    def __init__(self, value, demand_factor=None, demand_saving=None):
        self._value = value
        self._demand_factor_parameter = demand_factor
        self._demand_saving_parameter = demand_saving
    
    def value(self, ts, si):
        if self._demand_factor_parameter is None:
            demand_factor = 1.0
        else:
            demand_factor = self._demand_factor_parameter.value(ts, si)
        if self._demand_saving_parameter is None:
            demand_saving_factor = 1.0
        else:
            demand_saving_factor = self._demand_saving_parameter.factor(ts, si)
        return self._value * demand_factor * demand_saving_factor
parameter_registry.add(DemandParameter)
```

I think we need something between the two. It hadn't occured to me that the demand saving factor might vary in time - I can think of one project where this was indeed the case. Should be limit it to monthly, or just make `self.factors` a list of `Parameters`?

**Comment by jetuk** [comment]
By doing it as a numpy array of shape (levels, 12) or (levels, 366) you leave open a more efficient Cython optimisation process later. It's unlikely to have one level at monthly and another at daily, even if it was you could convert the monthly to daily anyway.

**Comment by jetuk** [comment]
I was just going to do this, but I have realised that in Cython `Parameter.value` is typed to return `double`. I would want to modify the `DailyProfileParameter` and `MonthlyProfileParameter` to return an array of doubles. They would then work inside a new control curve subclass as I wrote above.

**Comment by jetuk** [comment]
Right, so I've made some progress on this by making `BaseControlCurveParameter` accept multiple control curves. `ControlCurvePiecewiseParameter` now returns the current passing control curve number (zero if first passes, one if second, etc.) by default. You can tell it to return specific numbers at level instead to retain its current functionality. This is fine, and quite useful in and of itself.

I've also refactored `ParameterCollection` to `AggregatedParameter` which follows a similar approach to `AggregatedRecorder`.

I have then written a `ScaledParameter` which takes scales another `Parameter`. 

Ultimately this results in creating a demand saving paramter that would look something like this JSON. It requires a special control curve which is a mash up of control curve and monthly profile. I've not written that yet.

What do you think, is this too ugly?

```
    data = {
        'type': 'scaled',
        'scale': 50.0,    # Base demand
        'parameter': {
            'type': 'aggregated',
            'agg_func': 'product',
            'parameters': [
                {
                    'type': 'monthlyprofile',
                    'values': [1.0]*12   # Annual profile
                },
                {
                    'type': 'controlcurvemonthlyprofile',
                    'control_curve': [0.8, 0.6],
                    'values': [[1.0]*12,    # If above 80% then no saving
                                   [0.9]*12,    # If above 60% then 90% of base demand
                                   [0.8]*12,    # Otherwise 80% of base demand
                                  ]
                }
            ]
        }
    }
```

**Comment by snorfalorpagus** [comment]
The data structure makes sense but I think we need to alias it to make it more readable (like our `Reservoir` class).

```
data = {
    'type': 'demand',
    'value': 50.0,
    'profile': { ... }
    'demand_saving': 'referencehere'
}
```

You'll need a reference to the reservoir(s) in the control curve parameter. I assume you've already done this in the PR. I assume the control curve will be referenced rather than inline, as multiple demands will need to call it.

**Comment by snorfalorpagus** [comment]
The documentation for the control curves is slightly out of date now. e.g. `storage_node` is generally the first argument.

Also this needs fixing:
```
raise ValueError("Second dimension values must be size 12.".format(self._profile_size))
```

**Comment by jetuk** [comment]
Sorry, that is an error where?

**Comment by snorfalorpagus** [comment]
It's just a copy+paste error. Should be `{}` instead of `12`. I'm willing to make the remaining changes required to close this issue, but don't have time today - just making a note so I didn't forget.

**Comment by jetuk** [comment]
Sorry misread that. #157 expands the docstring for `ControlCurveParameter`
